jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -65,27 +76,69 @@ class S3TreeDirectoryNode(private val bucket: S3VirtualBucket, parent: S3TreeDir
             .contents()
             ?.filterNotNull()
             ?.filterNot { it.key() == key }
-            ?.map { S3TreeObjectNode(bucketName, this, it.key(), it.size(), it.lastModified()) as S3TreeNode }
+            ?.map { S3TreeObjectNode(bucket, this, it.key(), it.size(), it.lastModified()) as S3TreeNode }
             ?: emptyList()
 
         return (folders + s3Objects).sortedBy { it.key } + continuation
     }
-
-    fun removeAllChildren() {
-        cachedList = listOf()
-    }
 }
 
 private val fileTypeRegistry = FileTypeRegistry.getInstance()
 
-class S3TreeObjectNode(bucketName: String, parent: S3TreeDirectoryNode?, key: String, val size: Long, val lastModified: Instant) :
-    S3TreeNode(bucketName, parent, key) {
+open class S3TreeObjectNode(val bucket: S3VirtualBucket, parent: S3LazyLoadParentNode?, key: String, val size: Long, val lastModified: Instant) :
+    S3LazyLoadParentNode(bucket.name, parent, key) {
 
+    var showHistory: Boolean = false
+    var responseIterator: Iterator<ListObjectVersionsResponse>? = null
     private val fileType = fileTypeRegistry.getFileTypeByFileName(name)
 
     init {
         fileType.takeIf { it !is UnknownFileType }?.icon.let { icon = it }
     }
+
+    override fun loadObjects(continuationMarker: String?): List<S3TreeNode> {
+        if (showHistory) {
+            responseIterator = responseIterator ?: runBlocking {
+                bucket.listObjectVersionsPaginated(key)
+            }.iterator()
+
+            val nextPage = responseIterator
+                ?.next()
+                ?.versions()
+                ?.map { S3TreeObjectVersionNode(bucket, parent, key, it.size(), it.lastModified(), it.versionId()) as S3TreeNode }
+                ?: emptyList()
+
+            if (responseIterator?.hasNext() == true) {
+                return nextPage + S3TreeContinuationNode(
+                    bucketName,
+                    this,
+                    "${this.key}/${message("s3.load_more")}",
+                    (nextPage.last() as S3TreeObjectVersionNode).versionId
+                )
+            }
+            return nextPage
+        }
+        return emptyList()
+    }
+}
+
+class S3TreeObjectVersionNode(bucket: S3VirtualBucket, parent: S3LazyLoadParentNode?, key: String, size: Long, lastModified: Instant, val versionId: String) :
+    S3TreeObjectNode(bucket, parent, key, size, lastModified) {
+
+    override fun getName(): String {
```

Comment:
On second thought, it's a bit weird to have `<file>_<versionId>` in the tree when just the versionId would suffice (it's unclear that `getName` controls both the tree, filename output, and some other implicit stuff)

I'd probably add a case in `S3TreeCellRenderer` to handle this and we can revisit this later, unless @abrooksv has an opinion here

Reply:
I would treat only the versionID as the name here. See my other 2 comments in this class to help achieve that.

Reply:
Cool, I dropped the `<file>_` but kept extension, IMO it will make things easier when downloading multiple versions to download them with extension inherited from parent object.

Reply:
This seems like a fine compromise, may revisit in the future after some refactors to make less stuff relying on getName()